### ISSUE TITLE
Some fields insert unnecessary markup into templates

### ISSUE THREAD

**Issue by JoshBarr** [wagtail/wagtail#1214]
Current:

``` html
<div class="rich_text"><p>hello world</p></div>
```

Ideal:

``` html
<p>hello world</p>
```

**Comment by timheap** [comment]
This also happens with StreamBlocks, where each item is wrapped in a `<div class="block-text">`. I've yet to find a use for these blocks, and in fact have written code to *disable* them every time I use a StreamBlock.

**Comment by gasman** [comment]
I think this comes under the category of "it seemed like a good idea at the time" :-)

As I recall, the rationale for this was that you often want to give images (and potentially other elements) CSS rules such as `max-width: 100%` within rich text areas, without having to make those rules universal for all images across the site. The assumption was that the extra div wouldn't interfere with the styling for people who didn't want to make use of it (unless their CSS was particularly odd!) and therefore it would be better to include it as standard, rather than require template developers to insert a wrapper element manually when they did need one.

(I think it may also have been related to the ongoing difficulty in ensuring that rich-text content is consistently wrapped in `<p>` tags - the wrapper element gives an extra place to fix up margins.)

However, it does break our rule of Wagtail not dictating your front-end markup.

Clearly this isn't going to be straightforward to change, given that there's plenty of existing CSS in the wild that relies on that element. We'd probably want a flag in settings to restore the old behaviour. (Going through the formal deprecation process would be even more of a faff - having the flag set to the old behaviour by default, and then explicitly setting it to new behaviour in the settings file of the project template. That's probably overkill.)


My gut feeling is that the situation with StreamBlock is a bit different - I think that's more akin to Django forms, where you get a certain default rendering if you output it as a single unit (`{{ self.body }}`), and if you don't like that, you have the option to iterate over it instead. Admittedly, I can't think of many situations where the `<div class="block-foo">` wrapper is absolutely necessary to distinguish between different blocks for CSS purposes - if you're working with StructBlocks then you'll be specifying your own template (which can include all the identifying element classes you want), so it's only relevant if you have, say, two different 'flavours' of RichTextBlock.

It's worth noting that if we make both of these proposed changes, then in the default rendering, there will be no difference in markup between a single RichTextBlock containing two paragraphs, and two consecutive RichTextBlocks with one paragraph each. I can't decide whether that's a good thing or not!

**Comment by davecranwell** [comment]
In my implementations of streamfield so far i've never used the default rendering, so the difference between two individual paragraphs blocks and one two-paragraph block hasn't been an issue. As you say like with Django forms, if you're making even slightly custom layouts you start by discarding the default rendering.

I think with improved documentation we could persuade more people to use custom streamfield block rendering. Documenting with a direct comparison to django forms (if we aren't already) would provide much of the.

**Comment by JoshBarr** [comment]
For now I'll add an optional param to the richtext filter to disable the wrapping div. quick fix!

**Comment by MichielBijl** [comment]
Is there any way to disable the creation of these wrapping elements? It completely breaks my CSS :( I tend to use logical CSS instead of heaps of classes; so this breaks stuff like `section + section { … }`. Any help would be appreciated!

**Comment by SalahAdDin** [comment]
:+1: for solve this, i have very anoying thins when i was trying update hallo.js

**Comment by noahmanger** [comment]
Just wanted to +1 this. Is there a work-around to at least remove the `richtext` class?

**Comment by lb-** [comment]
@gasman should we put this back on the radar in the context of 2.0 / new draftail editor?

I personally ran into this issue yesterday with a StreamField RichTextBlock using specific layout rendering, also this question came up on Stack Overflow: 

https://stackoverflow.com/questions/47415978/custom-render-richtextblock-to-remove-div-class-rich-text

I am happy to do this as it seems simple, except not sure how to go about backwards compatibility (or if we want that for this area in 2.0).

**Comment by thibaudcolas** [comment]
We created https://github.com/springload/wagtaildraftail (a third-party plugin integrating https://springload.github.io/draftail/ into Wagtail) in part to address this.

This integration of Draftail gives developers more control over how the rich text HTML is rendered, so that classes can be added on arbitrary elements if needed. This makes the wrapper `.rich_text` div unnecessary.

Here is an [example from the config](https://github.com/springload/wagtaildraftail/blob/87f1ae3ade493c00daff021394051aa656136c10/tests/testapp/testapp/grains/rich_text.py#L130-L145], adding classes to lists to facilitate their styling:

```python
DRAFT_EXPORTER_BLOCK_MAP = dict(BLOCK_MAP, **{
    BLOCK_TYPES.UNORDERED_LIST_ITEM: {
        'element': 'li',
        'wrapper': 'ul',
        'wrapper_props': {'class': 'list-styled'},
    },
    BLOCK_TYPES.ORDERED_LIST_ITEM: {
        'element': 'li',
        'wrapper': 'ol',
        'wrapper_props': {'class': 'list-numbered'},
    },
})
```

If anyone is interested in this, a rough but good demonstrator of how this works is available at http://draftjs-exporter.herokuapp.com/. That one doesn't use WagtailDraftail, but is nonetheless a good demo of how this handles rich text.

------

@lb- for Wagtail 2.0, the general direction we discussed at the Bristol sprint was to focus on changing the rich text editor widget (so, Hallo to Draftail) first and foremost, then attack other issues that this will help address in following releases.

I'm not sure if/when we will get to porting the WagtailDraftail feature that allows customisation of how rich text HTML is rendered, but if/when that happens the migration path is rather simple: remove the `.rich_text` tag, and ask people to either add it manually in their templates, or configure this HTML generation to add the wrapper where appropriate.

**Comment by mikemonteith** [comment]
This is causing us some headaches. Would you accept a pull-request adding a new setting to disable the `<div class="rich-text">` divs?

My only solution at the moment is to monkey-patch the `RichText.__html__ method` 😟

**Comment by thibaudcolas** [comment]
Generally we want to keep the proliferation of settings in check – but we’ve recently added a setting for a very similar type of "Wagtail shouldn’t be opinionated about HTML" in [`WAGTAILEMBEDS_RESPONSIVE_HTML`](https://docs.wagtail.io/en/stable/reference/settings.html#embeds), making this type of front-end customization opt-in. So – it feels like that wrapper `div` could get the same treatment.

I’ll mark this as "Design decision needed" so we bring it up and discuss at a future core team meeting.

**Comment by allcaps** [comment]
At the moment the `richtext` filter hardcodes the wrapping div. 
If Wagtail used a template it would be easy to override and drop the div.

wagtail.core.templatetags.wagtailcore_tags.richtext returns:

     return mark_safe('<div class="rich-text">' + html + '</div>')

If we change that to:

     from django.template.loader import render_to_string
     ...
     return render_to_string('wagtailcore/richtext.html', {'html': html})

With a template at `wagtail/core/templates/wagtailcore/richtext.html`:

    <div class="rich-text">{{ html|safe }}</div>

Now an integrator can override by creating a template file in `/path/to/app/templates/wagtailcore/richtext.html`. Eg:

     {{ html|safe }}

**Comment by lb-** [comment]
Resolved via #6036

**Comment by rustybox** [comment]
Is there a timeline on 2.10?
